util/include/thread_pool.hpp

```diff
@@ -0,0 +1,115 @@
+// Concord
+//
+// Copyright (c) 2020 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the
+// LICENSE file.
+
+// Based on Mastering the C++17 STL, page 204:
+// https://books.google.bg/books?id=zJlGDwAAQBAJ&pg=PA205&lpg=PA205#
+
+#include <assertUtils.hpp>
+
+#include <condition_variable>
+#include <future>
+#include <mutex>
+#include <queue>
+#include <thread>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace concord::util {
+
+// A thread pool that supports any callable object with any return type. Returns std::future objects to users.
+class ThreadPool {
+ public:
+  // Starts the thread pool with thread_count > 0 threads.
+  ThreadPool(unsigned int thread_count) noexcept {
+    Assert(thread_count > 0);
+    for (auto i = 0u; i < thread_count; ++i) {
+      threads_.emplace_back([this]() { loop(); });
+    }
+  }
+
+  // Starts the thread pool with the maximum number of concurrent threads supported by the implementation.
+  ThreadPool() noexcept
+      : ThreadPool{std::thread::hardware_concurrency() > 0 ? std::thread::hardware_concurrency() : 1} {}
+
+  // Stops the thread pool. Waits for the currently executing tasks only (will not exhaust the queues).
+  ~ThreadPool() noexcept {
+    {
+      auto lock = std::lock_guard{task_queue_.mutex};
+      task_queue_.stop = true;
+    }
+    task_queue_.cv.notify_all();
+    for (auto& t : threads_) {
+      t.join();
+    }
+  }
+
+ public:
+  // Executes the passed function (or any callable) in a pool thread. Returns a future to the result.
+  // Note: The returned future's destructor doesn't block if the future value hasn't been retrieved. This is in contrast
+  // to the futures returned by std::async that block.
+  template <class F>
+  auto async(F&& func) {
+    using ResultType = std::invoke_result_t<std::decay_t<F>>;
```

Comment:
I've tried, but still don't fully understand why `std::decay_t<F>` is needed here. Do function pointers not work without it? What callable needs to be decayed?

Reply:
It's a bit tricky and not extremely practical: https://stackoverflow.com/questions/37834567/decay-types-before-passing-to-stdresult-of

Reply:
Thanks!